File: optimized.py
import collections
import sys

counts = collections.Counter()
remaining = ''
while True:
    chunk = sys.stdin.read(64*1024)
    if not chunk:
        break
    chunk = remaining + chunk
    last_lf = chunk.rfind('\n')  # process to last LF character
    if last_lf == -1:
        remaining = ''
    else:
        remaining = chunk[last_lf+1:]
        chunk = chunk[:last_lf]
    counts.update(chunk.lower().split())

counts.update(remaining.lower().split())

for word, count in counts.most_common():
    print(word, count)

File: optimized.rb
counts = Hash.new(0)
bufsize = 64*1024
remaining = ''
while 1 do
  chunk = STDIN.read(bufsize)
  if !chunk
    break
  end
  chunk = remaining + chunk
  last_lf = chunk.rindex("\n")
  if last_lf.nil?
    remaining = ''
  else
    remaining = chunk[last_lf+1..]
    chunk = chunk[..last_lf]
  end
  chunk.downcase.split.each{ |w| counts[w] += 1 }
end

counts.sort_by{|k,v| -v}.each{|k,v| puts "#{k} #{v}"}
